File: packages/dotnet/BTCPayServer.Plugins.OpenReceive/Nwc/Settlement.cs
#nullable enable
using System;

namespace BTCPayServer.Plugins.OpenReceive.Nwc;

/// <summary>What one wallet row says about its invoice: the C# twin of core's <c>TransactionSettlementDetection</c>.</summary>
/// <param name="Status">pending, settled, expired or failed.</param>
/// <param name="FinalitySignal">settled_at, transaction_state or state — the field that decided settlement.</param>
public sealed record SettlementDetection(bool Settled, string Status, string? FinalitySignal, bool PreimagePresent, long? SettledAt);

/// <summary>
/// The settlement rule — the C# twin of <c>isTransactionSettled</c> (core/nwc/client.ts) and
/// <c>classifyTransactionSettlement</c> (core/settlement/index.ts). A row is settled iff
/// <c>settled_at &gt; 0</c> or either state key says settled (case-insensitively); a preimage
/// alone never settles.
/// </summary>
public static class Settlement
{
    public static bool IsState(string? value, string expected) =>
        value is not null && string.Equals(value, expected, StringComparison.OrdinalIgnoreCase);

    public static bool IsSettled(NwcTransaction transaction) =>
        transaction.SettledAt is > 0
        || IsState(transaction.TransactionState, "settled")
        || IsState(transaction.State, "settled");

    /// <summary>Same field precedence as <see cref="IsSettled"/>, so the reported signal is the one that fired.</summary>
    public static string? FinalitySignal(NwcTransaction transaction)
    {
        if (!IsSettled(transaction)) return null;
        if (transaction.SettledAt is > 0) return "settled_at";
        if (IsState(transaction.TransactionState, "settled")) return "transaction_state";
        if (IsState(transaction.State, "settled")) return "state";
        return null;
    }

    public static bool IsExpired(NwcTransaction transaction) =>
        IsState(transaction.State, "expired") || IsState(transaction.TransactionState, "expired");

    public static bool IsFailed(NwcTransaction transaction) =>
        IsState(transaction.State, "failed") || IsState(transaction.TransactionState, "failed");

    public static SettlementDetection Classify(NwcTransaction transaction)
    {
        var signal = FinalitySignal(transaction);
        var status = signal is not null ? "settled"
            : IsExpired(transaction) ? "expired"
            : IsFailed(transaction) ? "failed"
            : "pending";
        return new SettlementDetection(
            Settled: status == "settled",
            Status: status,
            FinalitySignal: signal,
            PreimagePresent: transaction.Preimage is not null,
            SettledAt: transaction.SettledAt);
    }
}

File: packages/python/openreceive/src/openreceive/settlement.py
"""The settlement classification rule (`settlement-detection` vectors):
settled requires `settled_at > 0` or a settled state; a preimage alone is
corroborating evidence, never proof. Twin of Ruby `OpenReceive::Settlement`.
"""

from __future__ import annotations

from typing import Any

from openreceive.values import stringify


def _state_is(data: dict[str, Any], expected: str) -> bool:
    # Raw wallet states compare case-insensitively (the JS isTransactionState rule).
    return any(
        isinstance(value, str) and value.lower() == expected
        for value in (data.get("state"), data.get("transaction_state"))
    )


def _settled_at_positive(value: object) -> bool:
    if isinstance(value, bool):
        return False
    if isinstance(value, (int, float)):
        return value > 0
    if isinstance(value, str):
        try:
            return int(value.strip()) > 0
        except ValueError:
            return False
    return False


def is_settled(transaction: object) -> bool:
    data = stringify(transaction)
    return _settled_at_positive(data.get("settled_at")) or _state_is(data, "settled")


def status(transaction: object) -> str:
    """One of settled | expired | failed | pending."""
    data = stringify(transaction)
    if is_settled(data):
        return "settled"
    if _state_is(data, "expired"):
        return "expired"
    if _state_is(data, "failed"):
        return "failed"
    return "pending"
